sdk/core/azure-core/inc/azure/core/nullable.hpp

```diff
@@ -261,10 +261,72 @@ namespace Azure { namespace Core {
       return std::move(m_value);
     }
 
+    // observers
+
     /**
      * @brief `operator bool` on the condition of #Azure::Core::Nullable::HasValue.
      */
-    explicit operator bool() const noexcept { return HasValue(); }
+    constexpr explicit operator bool() const noexcept { return HasValue(); }
+
+    /**
+     * @brief Accesses the contained value.
+     * @return Returns a pointer to the contained value
+     * @warning The behavior is undefined if `*this` does not contain a value.
+     * @note This operator does not check whether the optional contains a value!
+             You can do so manually by using #HasValue() or simply operator #bool().
+             Alternatively, if checked access is needed, #GetValue() or #ValueOr() may be used.
+     */
+    constexpr const T* operator->() const { return std::addressof(m_value); }
+
+    /**
+     * @brief Accesses the contained value.
+     * @return Returns a pointer to the contained value
+     * @warning The behavior is undefined if `*this` does not contain a value.
+     * @note This operator does not check whether the optional contains a value!
+             You can do so manually by using #HasValue() or simply operator #bool().
+             Alternatively, if checked access is needed, #GetValue() or #ValueOr() may be used.
+     */
+    constexpr T* operator->() { return _STD addressof(m_value); }
+
+    /**
+     * @brief Accesses the contained value.
+     * @return Returns a reference to the contained value
+     * @warning The behavior is undefined if `*this` does not contain a value.
+     * @note This operator does not check whether the optional contains a value!
+             You can do so manually by using #HasValue() or simply operator #bool().
+             Alternatively, if checked access is needed, #GetValue() or #ValueOr() may be used.
+     */
+    constexpr const T& operator*() const& { return m_value; }
+
+    /**
+     * @brief Accesses the contained value.
+     * @return Returns a reference to the contained value
+     * @warning The behavior is undefined if `*this` does not contain a value.
+     * @note This operator does not check whether the optional contains a value!
+             You can do so manually by using #HasValue() or simply operator #bool().
+             Alternatively, if checked access is needed, #GetValue() or #ValueOr() may be used.
+     */
+    constexpr T& operator*() & { return m_value; }
+
+    /**
+     * @brief Accesses the contained value.
+     * @return Returns a reference to the contained value
+     * @warning The behavior is undefined if `*this` does not contain a value.
+     * @note This operator does not check whether the optional contains a value!
+             You can do so manually by using #HasValue() or simply operator #bool().
+             Alternatively, if checked access is needed, #GetValue() or #ValueOr() may be used.
+     */
+    constexpr T&& operator*() && { return std::move(m_value); }
```

Comment:
How does the caller know they are calling this "extracting" operator rather than the other `*` operator?
Should this be an explicit named method like `ExtractValue`, following consistency with other such APIs within the SDK?

I would love to understand more what the subtle differences between all these operators and how the caller sees what's happening between each.

Reply:
Looks like we already have this with `GetValue()`.

Reply:
The difference is in the check.  The operators allow you to interact with the type with validation responsibilities belonging to the user.  GetValue() has perf overhead because it does additional checks.

Reply:
the `std::move(m_value)` is letting `m_value` as a moved reference.  That means using `*` operator would work only once to move construct `T` out of the `nullable<T>`

Reply:
Add a test like
```cpp
Nullable<std::string> n = "some value";

std::cout << *n << std::endl;
std::cout << *n << std::endl;  // I fear this won't print the value any more because it was move on previous line

```